### PR TITLE
Feature/add tensorrt backend

### PR DESCRIPTION
**Pull request by emmadrigal**

### CODE REVIEW

File: examples/r2i/tensorrt/Makefile.am [hunk 1]
```diff
@@ -0,0 +1,35 @@
+# Copyright (C) 2018 RidgeRun, LLC (http://www.ridgerun.com)
```
Comment (michaelgruner) [inline examples/r2i/tensorrt/Makefile.am]: Update headers 2018-2020

File: configure.ac [hunk 2]
```diff
@@ -52,8 +52,13 @@ RR_CHECK_FEATURE_LIB(TENSORFLOW, TensorFlow Installation,
 AC_LANG_PUSH([C++])
 RR_CHECK_FEATURE_LIB(TFLITE, TensorFlow lite Installation,
                      tensorflow-lite, TfLiteTypeGetName, tensorflow/lite/model.h, no)
+
+RR_CHECK_FEATURE_LIB(TENSORRT, TensorRT Installation,
+                     nvinfer, createInferBuilder_INTERNAL, NvInfer.h, no)
 AC_LANG_POP([C++])
 
+AM_COND_IF([HAVE_TENSORRT], [RR_CHECK_CUDA], [])
```
Comment (michaelgruner) [inline configure.ac]: Is this really needed?

File: r2i/iframe.h [hunk 3]
```diff
@@ -65,6 +67,12 @@ class IFrame {
    */
   virtual ImageFormat GetFormat () = 0;
 
+  /**
+   * \brief Gets the Image data type set to the Frame.
+   * \return Image data type.
+   */
+  virtual DataType GetDataType () { return r2i::DataType::Id::UNKNOWN_DATATYPE; }
```
Comment (michaelgruner) [inline r2i/iframe.h]: I'd rather keep all methods abstract.

File: r2i/ncsdk/frame.h [hunk 4]
```diff
@@ -20,7 +20,8 @@ namespace ncsdk {
 class Frame : public IFrame {
  public:
   RuntimeError Configure (void *in_data, int width,
-                          int height, r2i::ImageFormat::Id format) override;
+                          int height, r2i::ImageFormat::Id format,
+                          r2i::DataType::Id data_type = r2i::DataType::Id::UNKNOWN_DATATYPE) override;
```
Comment (michaelgruner) [inline r2i/ncsdk/frame.h]: This feels incorrect. If now one is using the interface we should not have a common interface for everyone.

File: r2i/tensorrt/Makefile.am [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+    $(RR_CFLAGS)                 \
+    $(TENSORRT_CFLAGS)           \
+    $(CODE_COVERAGE_CFLAGS)      \
+    -I$(CUDA_INCDIR)
+
+libtensorrt_la_CXXFLAGS =        \
+    $(RR_CXXFLAGS)               \
+    $(TENSORRT_CFLAGS)           \
+    $(CODE_COVERAGE_CXXFLAGS)    \
+    $(CUDA_LIBS)
```
Comment (michaelgruner) [inline r2i/tensorrt/Makefile.am]: ?

File: r2i/tensorrt/Makefile.am [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+
+libtensorrt_la_CPPFLAGS =        \
+    $(RR_CPPFLAGS)               \
+    $(CODE_COVERAGE_CPPFLAGS)    \
+    -I$(CUDA_INCDIR)
+
+libtensorrt_la_CFLAGS =          \
+    $(RR_CFLAGS)                 \
+    $(TENSORRT_CFLAGS)           \
+    $(CODE_COVERAGE_CFLAGS)      \
```
Comment (michaelgruner) [inline r2i/tensorrt/Makefile.am]: Remove duplicate flags (the ones cppflags already have)

File: r2i/tensorrt/engine.cc [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+
+  error.Clean ();
+
+  auto prediction = std::make_shared<Prediction>();
+
+  std::vector < void *> buffers;
+
+  if (cuda_engine->getNbBindings () != 2) {
+    error.Set (RuntimeError::Code::INCOMPATIBLE_MODEL,
+               "Unable to run prediction on model");
```
Comment (michaelgruner) [inline r2i/tensorrt/engine.cc]: Write a more descriptive message

File: r2i/tensorrt/engine.cc [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+r2i::RuntimeError Engine::SetBatchSize (const int batch_size) {
+  r2i::RuntimeError error;
+  if (this->model == nullptr) {
+    error.Set (RuntimeError::Code::WRONG_API_USAGE,
+               "Model must have been added before setting a batch size");
+    return error;
+  }
+  if (this->model->GetTRCudaEngine()->getMaxBatchSize() < batch_size) {
+    error.Set (RuntimeError::Code::WRONG_API_USAGE,
+               "Batch size can't be larger than the one used to generate the engine");
```
Comment (michaelgruner) [inline r2i/tensorrt/engine.cc]: which is?

File: r2i/tensorrt/engine.h [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+#ifndef R2I_TENSORRT_ENGINE_H
+#define R2I_TENSORRT_ENGINE_H
+
+#include <r2i/iengine.h>
+
+#include <memory>
+
+#include <r2i/tensorrt/model.h>
+
+#define RAM_ARRAY_SIZE 11
```
Comment (michaelgruner) [inline r2i/tensorrt/engine.h]: what is this?

File: r2i/tensorrt/engine.cc [hunk 10]
```diff
@@ -0,0 +1,168 @@
...
+ *
+ * The contents of this software are proprietary and confidential to RidgeRun,
+ * LLC.  No part of this program may be photocopied, reproduced or translated
+ * into another programming language without prior written consent of
+ * RidgeRun, LLC.  The user is free to modify the source code after obtaining
+ * a software license from RidgeRun.  All source code changes must be provided
+ * back to RidgeRun without any encumbrance.
+ */
+
+#include "r2i/tensorrt/engine.h"
```
Comment (michaelgruner) [inline r2i/tensorrt/engine.cc]: Order headers

File: r2i/tensorrt/engine.cc [hunk 11]
```diff
@@ -0,0 +1,168 @@
...
+#include "r2i/tensorrt/frame.h"
+
+#include <cuda.h>
+#include <cuda_runtime.h>
+#include <NvInfer.h>
+#include "NvInferPlugin.h"
+#include <vector>
+
+static int
+getSizeFromType (nvinfer1::DataType type) {
```
Comment (michaelgruner) [inline r2i/tensorrt/engine.cc]: I feel this is a duplicate with our new datatype class. Can we convert nvinfer::DataType into r2i::DataType and get the size from there?

File: r2i/tensorrt/frame.cc [hunk 12]
```diff
@@ -0,0 +1,102 @@
...
+  this->frame_data = std::shared_ptr <void> (buff, cudaFree);
+
+  cuda_error = cudaMemcpy(this->frame_data.get(), in_data, frame_size,
+                          cudaMemcpyHostToDevice);
+  if (cudaSuccess != cuda_error) {
+    error.Set (RuntimeError::Code::MEMORY_ERROR,
+               "Unable to copy data to buffer");
+    return error;
+  }
+
```
Comment (michaelgruner) [inline r2i/tensorrt/frame.cc]: Don't we need to sync here?

File: r2i/tensorrt/frame.h [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+
+ private:
+  /* This backend stores its own copy of */
+  std::shared_ptr<void> frame_data;
+  int frame_width;
+  int frame_height;
+  ImageFormat frame_format;
+  DataType data_type;
+
+  RuntimeError Validate (int64_t dims[], int64_t num_dims);
```
Comment (michaelgruner) [inline r2i/tensorrt/frame.h]: Where is this used?

File: r2i/tensorrt/frameworkfactory.cc [hunk 14]
```diff
@@ -0,0 +1,64 @@
...
+  error.Clean ();
+
+  return std::unique_ptr<IFrame> (new Frame);
+}
+
+r2i::FrameworkMeta FrameworkFactory::GetDescription (RuntimeError &error) {
+  const FrameworkMeta meta {
+    .code = r2i::FrameworkCode::TENSORRT,
+    .name = "TensorRT",
+    .description = "NVidia's Tensorrt",
```
Comment (michaelgruner) [inline r2i/tensorrt/frameworkfactory.cc]: Choose one! TensorRT or Tensorrt

File: r2i/tensorrt/frameworkfactory.cc [hunk 15]
```diff
@@ -0,0 +1,64 @@
...
+    .version = "7.0.0" // TODO get version programatically
```
Comment (michaelgruner) [inline r2i/tensorrt/frameworkfactory.cc]: Please fix this for this release

File: r2i/tensorrt/loader.cc [hunk 16]
```diff
@@ -0,0 +1,112 @@
...
+
+#include <fstream>
+#include <iostream>
+#include <vector>
+
+#include "r2i/imodel.h"
+#include "r2i/tensorrt/model.h"
+
+static void iRuntimeDeleter (nvinfer1::IRuntime *p) {
+  if (p)
```
Comment (michaelgruner) [inline r2i/tensorrt/loader.cc]: I recommend making the muscle memory to always use brackets even for one line ifs. This is one of the most common bug sources.

File: r2i/tensorrt/loader.cc [hunk 17]
```diff
@@ -0,0 +1,112 @@
...
+    p->destroy ();
+}
+
+static void ICudaEngineDeleter (nvinfer1::ICudaEngine *p) {
+  if (p)
+    p->destroy ();
+}
+
+static void IExecutionContextDeleter (nvinfer1::IExecutionContext *p) {
```
Comment (michaelgruner) [inline r2i/tensorrt/loader.cc]: Convert these three into a template.

File: r2i/tensorrt/loader.cc [hunk 18]
```diff
@@ -0,0 +1,112 @@
...
+    error.Set (RuntimeError::Code::INCOMPATIBLE_MODEL,
+               "Unable to load cached engine");
+    return nullptr;
+  }
+
+  std::shared_ptr<r2i::tensorrt::Model> model = std::make_shared<Model>();
+
+  error = model->SetContext(context);
+  if (error.IsError ()) {
+    model = nullptr;
```
Comment (michaelgruner) [inline r2i/tensorrt/loader.cc]: What about the error here?

File: r2i/tensorrt/loader.cc [hunk 19]
```diff
@@ -0,0 +1,112 @@
...
+  } else {
+    error = model->SetCudaEngine(engine);
+    if (error.IsError ()) {
+      model = nullptr;
```
Comment (michaelgruner) [inline r2i/tensorrt/loader.cc]: What about the error here

File: r2i/tensorrt/loader.h [hunk 20]
```diff
@@ -0,0 +1,33 @@
...
+
+namespace r2i {
+namespace tensorrt {
+
+class Loader : public ILoader {
+ public:
+  virtual std::shared_ptr<r2i::IModel> Load (const std::string &in_path,
+      r2i::RuntimeError &error) override;
+ private:
+  std::shared_ptr<Model> model;
```
Comment (michaelgruner) [inline r2i/tensorrt/loader.h]: Why do we need to save the model?

File: r2i/tensorrt/model.h [hunk 21]
```diff
@@ -0,0 +1,82 @@
...
+          Severity::kINFO:
+        tag = "INFO";
+        break;
+      case
+          Severity::kVERBOSE:
+        tag = "VERBOSE";
+        break;
+    }
+
+    std::cerr << "[RR]" << "[" << tag << "] " << std::string (msg) << std::endl;
```
Comment (michaelgruner) [inline r2i/tensorrt/model.h]: We don't want R2I printing to console. We need to be able to configure a Logger.

File: r2i/tensorrt/parameters.h [hunk 22]
```diff
@@ -0,0 +1,131 @@
...
+    Accessor (Parameters *target) : target(target) {}
+    virtual RuntimeError Set () = 0;
+    virtual RuntimeError Get () = 0;
+    virtual ~Accessor() {}
+
+   public:
+    Parameters *target;
+  };
+
+  class StringAccessor : public Accessor {
```
Comment (michaelgruner) [inline r2i/tensorrt/parameters.h]: Do we need all these in the header?

File: r2i/tensorrt/prediction.cc [hunk 23]
```diff
@@ -0,0 +1,99 @@
...
+  cuda_error = cudaMemcpy(buff.get(),
+                          result_buffer.get(),
+                          size,
+                          cudaMemcpyDeviceToHost);
+  if (cudaSuccess != cuda_error) {
+    error.Set (RuntimeError::Code::MEMORY_ERROR,
+               "Unable to read data from CUDA");
+    return error;
+  }
+
```
Comment (michaelgruner) [inline r2i/tensorrt/prediction.cc]: Do we need to sync here?

File: tests/unit/r2i/tensorrt/engine.cc [hunk 24]
```diff
@@ -0,0 +1,163 @@
...
+ * The contents of this software are proprietary and confidential to RidgeRun,
+ * LLC.  No part of this program may be photocopied, reproduced or translated
+ * into another programming language without prior written consent of
+ * RidgeRun, LLC.  The user is free to modify the source code after obtaining
+ * a software license from RidgeRun.  All source code changes must be provided
+ * back to RidgeRun without any encumbrance.
+ */
+
+#include <r2i/r2i.h>
+#include <r2i/tensorrt/engine.h>
```
Comment (michaelgruner) [inline tests/unit/r2i/tensorrt/engine.cc]: Why are we including these and not the interfaces? Can't we test classes independently?

File: tests/unit/r2i/tensorrt/engine.cc [hunk 25]
```diff
@@ -0,0 +1,163 @@
...
+TEST (TensorRTEngine, SetModelNull) {
+  error = engine.SetModel (nullptr);
+  LONGS_EQUAL (r2i::RuntimeError::Code::NULL_PARAMETER, error.GetCode ());
+}
+TEST (TensorRTEngine, SetModelInvalid) {
+  error = engine.SetModel (inc_model);
+  LONGS_EQUAL (r2i::RuntimeError::Code::INCOMPATIBLE_MODEL, error.GetCode ());
+}
+
+TEST (TensorRTEngine, StartEngine) {
```
Comment (michaelgruner) [inline tests/unit/r2i/tensorrt/engine.cc]: Failed start?

File: tests/unit/r2i/tensorrt/engine.cc [hunk 26]
```diff
@@ -0,0 +1,163 @@
...
+  error = engine.Start ();
+  LONGS_EQUAL (r2i::RuntimeError::Code::EOK, error.GetCode ());
+}
+
+TEST (TensorRTEngine, StopEngine) {
```
Comment (michaelgruner) [inline tests/unit/r2i/tensorrt/engine.cc]: Failed stop?

File: tests/unit/r2i/tensorrt/engine.cc [hunk 27]
```diff
@@ -0,0 +1,163 @@
...
+
+TEST (TensorRTEngine, SetBatchSizeSucess) {
+  error = engine.SetModel (model);
+  LONGS_EQUAL (r2i::RuntimeError::Code::EOK, error.GetCode ());
+
+  error = engine.SetBatchSize (MAX_BATCH_SIZE);
+  LONGS_EQUAL (r2i::RuntimeError::Code::EOK, error.GetCode ());
+}
+
+TEST (TensorRTEngine, PredictExecutionError) {
```
Comment (michaelgruner) [inline tests/unit/r2i/tensorrt/engine.cc]: Predict null parameters

File: tests/unit/r2i/tensorrt/loader.cc [hunk 28]
```diff
@@ -0,0 +1,122 @@
...
+
+TEST (TensorRTLoader, LoadInvalidFile) {
+  incompatible_model = true;
+
+  auto model = loader.Load(__FILE__, error);
+
+  LONGS_EQUAL (r2i::RuntimeError::Code::INCOMPATIBLE_MODEL, error.GetCode ());
+}
+
+TEST (TensorRTLoader, BadChechedEngine) {
```
Comment (michaelgruner) [inline tests/unit/r2i/tensorrt/loader.cc]: typo

File: tests/unit/r2i/tensorrt/mockruntime.cc [hunk 29]
```diff
@@ -0,0 +1,63 @@
+// namespace mock {
```
Comment (michaelgruner) [inline tests/unit/r2i/tensorrt/mockruntime.cc]: Erase dirty code

File: tests/unit/r2i/tensorrt/frame.cc [hunk 30]
```diff
@@ -0,0 +1,185 @@
...
+#include <cuda_runtime.h>
+
+#define FRAME_WIDTH  2
+#define FRAME_HEIGHT  2
+
+bool cudaMallocError = false;
+bool cudaMemCpyError = false;
+size_t cudaRequestedSize = 0;
+
+__host__ __cudart_builtin__ cudaError_t CUDARTAPI
```
Comment (michaelgruner) [inline tests/unit/r2i/tensorrt/frame.cc]: How does this compile? doesn't this need to be interpreted by NVCC?

File: examples/r2i/tensorrt/stb_image_resize.h [hunk 31]
```diff
@@ -0,0 +1,2746 @@
+/* stb_image_resize - v0.94 - public domain image resizing
```
Comment (rrcarlosrodriguez) [inline examples/r2i/tensorrt/stb_image_resize.h]: This file is duplicated in tensorflow, tflite and tensorrt. We should refactor it!

File: examples/r2i/tensorrt/stb_image.h [hunk 32]
```diff
@@ -0,0 +1,8419 @@
+/* stb_image - v2.15 - public domain image loader - http://nothings.org/stb_image.h
```
Comment (rrcarlosrodriguez) [inline examples/r2i/tensorrt/stb_image.h]: This file is duplicated in tensorflow, tflite and tensorrt. We should refactor it!

File: examples/r2i/tensorrt/yolov2.cc [hunk 33]
```diff
@@ -0,0 +1,261 @@
...
+ * RidgeRun, LLC.  The user is free to modify the source code after obtaining
+ * a software license from RidgeRun.  All source code changes must be provided
+ * back to RidgeRun without any encumbrance.
+*/
+
+#include <getopt.h>
+#include <iostream>
+#include <memory>
+#include <string>
+#include <r2i/r2i.h>
```
Comment (rrcarlosrodriguez) [inline examples/r2i/tensorrt/yolov2.cc]: Alphabetical order

File: examples/r2i/tensorrt/yolov2.cc [hunk 34]
```diff
@@ -0,0 +1,261 @@
...
+
+static void
+box_to_pixels (BBox *normalized_box, int row, int col, int box) {
+  int grid_size = 32;
+  const float box_anchors[] =
+  { 1.08, 1.19, 3.42, 4.41, 6.63, 11.38, 9.42, 5.11, 16.62, 10.52 };
+
+  /* adjust the box center according to its cell and grid dim */
+  normalized_box->x = (col + sigmoid (normalized_box->x)) * grid_size;
+  normalized_box->y = (row + sigmoid (normalized_box->y)) * grid_size;
```
Comment (rrcarlosrodriguez) [inline examples/r2i/tensorrt/yolov2.cc]: Pointer guards

File: examples/r2i/tensorrt/yolov2.cc [hunk 35]
```diff
@@ -0,0 +1,261 @@
...
+
+std::unique_ptr<float[]> PreProcessImage (const unsigned char *input,
+    int width, int height, int reqwidth, int reqheight) {
+
+  const int channels = 3;
+  const int scaled_size = channels * reqwidth * reqheight;
+  std::unique_ptr<unsigned char[]> scaled (new unsigned char[scaled_size]);
+  std::unique_ptr<float[]> adjusted (new float[scaled_size]);
+
+  stbir_resize_uint8(input, width, height, 0, scaled.get(), reqwidth,
```
Comment (rrcarlosrodriguez) [inline examples/r2i/tensorrt/yolov2.cc]: Pointer guards

File: examples/r2i/tensorrt/yolov2.cc [hunk 36]
```diff
@@ -0,0 +1,261 @@
...
+  normalized_box->width =
+    pow (M_E, normalized_box->width) * box_anchors[2 * box] * grid_size;
+  normalized_box->height =
+    pow (M_E, normalized_box->height) * box_anchors[2 * box + 1] * grid_size;
+}
+
+static r2i::RuntimeError
+parse_boxes(std::shared_ptr<r2i::IPrediction> prediction) {
+  r2i::RuntimeError error;
+  float *network_output = static_cast<float *>(prediction->GetResultData());
```
Comment (rrcarlosrodriguez) [inline examples/r2i/tensorrt/yolov2.cc]: Pointer guard

File: r2i/datatype.cc [hunk 37]
```diff
@@ -0,0 +1,58 @@
+/* Copyright (C) 2020 RidgeRun, LLC (http://www.ridgerun.com)
```
Comment (rrcarlosrodriguez) [inline r2i/datatype.cc]: 2018-2020

File: r2i/datatype.h [hunk 38]
```diff
@@ -0,0 +1,98 @@
+/* Copyright (C) 2020 RidgeRun, LLC (http://www.ridgerun.com)
```
Comment (rrcarlosrodriguez) [inline r2i/datatype.h]: 2018-2020

File: r2i/datatype.h [hunk 39]
```diff
@@ -0,0 +1,98 @@
...
+#define R2I_DATATYPE_H
+
+#include <string>
+
+/**
+ * R2Inference Namespace
+ */
+namespace r2i {
+/**
+ *  Implements the Image formats handling class for r2i library.
```
Comment (rrcarlosrodriguez) [inline r2i/datatype.h]: Image formats? wrong comment

File: r2i/datatype.h [hunk 40]
```diff
@@ -0,0 +1,98 @@
...
+   */
+  int GetBytesPerPixel ();
+
+ private:
+  Id id;
+};
+
+}
+
+#endif // R2I_IMAGEFORMAT_H
```
Comment (rrcarlosrodriguez) [inline r2i/datatype.h]: IMAGEFORMAT? wrong comment

File: r2i/datatype.h [hunk 41]
```diff
@@ -0,0 +1,98 @@
...
+    /**
+     * Integer 8 bits per pixel
+     */
+    INT8,
+
+    UNKNOWN_DATATYPE,
+  };
+
+  /**
+   * \brief Creates and initializes a format to Unknown format.
```
Comment (rrcarlosrodriguez) [inline r2i/datatype.h]: Why do you use `format` everywhere? you should use `data type`

File: r2i/datatype.cc [hunk 42]
```diff
@@ -0,0 +1,58 @@
...
+
+#include "r2i/datatype.h"
+
+#include <string>
+#include <unordered_map>
+
+namespace r2i {
+
+static std::unordered_map<int, std::pair<const std::string, int>>
+format_descriptors ({
```
Comment (rrcarlosrodriguez) [inline r2i/datatype.cc]: data_type_descriptors would be a better name

File: r2i/frameworks.h [hunk 43]
```diff
@@ -36,6 +36,11 @@ enum FrameworkCode {
    */
   TFLITE,
 
+  /**
+   * NVidia's TensorFlow Lite
```
Comment (rrcarlosrodriguez) [inline r2i/frameworks.h]: `NVIDIA`, `NVidia` is not accepted

TensorFlow Lite???, TensorRT

File: r2i/iframe.h [hunk 44]
```diff
@@ -14,6 +14,7 @@
 
 #include <r2i/runtimeerror.h>
 #include <r2i/imageformat.h>
+#include <r2i/datatype.h>
```
Comment (rrcarlosrodriguez) [inline r2i/iframe.h]: Alphabetical order

File: r2i/iframeworkfactory.cc [hunk 45]
```diff
@@ -18,6 +18,7 @@
 #include "ncsdk/frameworkfactory.h"
 #include "tensorflow/frameworkfactory.h"
 #include "tflite/frameworkfactory.h"
+#include "tensorrt/frameworkfactory.h"
```
Comment (rrcarlosrodriguez) [inline r2i/iframeworkfactory.cc]: Alphabetical order

File: r2i/tensorrt/Makefile.am [hunk 46]
```diff
@@ -0,0 +1,59 @@
+# Copyright (C) 2018 RidgeRun, LLC (http://www.ridgerun.com)
```
Comment (rrcarlosrodriguez) [inline r2i/tensorrt/Makefile.am]: 2018-2020

File: r2i/tensorrt/Makefile.am (hunk 5 above)
Comment (rrcarlosrodriguez) [reply r2i/tensorrt/Makefile.am]: weird

File: r2i/tensorrt/engine.h [hunk 47]
```diff
@@ -0,0 +1,52 @@
...
+ * a software license from RidgeRun.  All source code changes must be provided
+ * back to RidgeRun without any encumbrance.
+*/
+
+#ifndef R2I_TENSORRT_ENGINE_H
+#define R2I_TENSORRT_ENGINE_H
+
+#include <r2i/iengine.h>
+
+#include <memory>
```
Comment (rrcarlosrodriguez) [inline r2i/tensorrt/engine.h]: Includes look weird. Order them

File: configure.ac (hunk 2 above)
Comment (emmadrigal) [reply configure.ac]: The TensorRT backend needs cuda headers and libraries in order to compile (it uses `cudaMalloc` & `cudaMemcpy`). This ensures that this is added for when setting up this backend

File: r2i/datatype.h (hunk 41 above)
Comment (emmadrigal) [reply r2i/datatype.h]: I based this on the imageformats class, I forgot to change the comments over

File: r2i/tensorrt/Makefile.am (hunk 5 above)
Comment (emmadrigal) [reply r2i/tensorrt/Makefile.am]: This is added by the `RR_CUDA_CHECK` and is needed in order to use `cudaMalloc` and `cudaMemcpu`

File: r2i/tensorrt/frame.cc (hunk 12 above)
Comment (emmadrigal) [reply r2i/tensorrt/frame.cc]: `CudaMemcpy` blocks by itself

File: r2i/tensorrt/prediction.cc (hunk 23 above)
Comment (emmadrigal) [reply r2i/tensorrt/prediction.cc]: `cudaMemcpy` forces the sync

File: tests/unit/r2i/tensorrt/engine.cc (hunk 25 above)
Comment (emmadrigal) [reply tests/unit/r2i/tensorrt/engine.cc]: This engine doesn't have a start logic, this test is only for the interface implemetation, see [here](https://github.com/RidgeRun/r2inference/blob/cc4c5646924b4706c60cc0159275aa61d643b393/r2i/tensorrt/engine.cc#L67)

File: tests/unit/r2i/tensorrt/frame.cc (hunk 30 above)
Comment (emmadrigal) [reply tests/unit/r2i/tensorrt/frame.cc]: All of these flags are already included by the cuda headers, the actual function is meant to be run on the Host side and therefore compilable by gcc

File: tests/unit/r2i/tensorrt/engine.cc (hunk 24 above)
Comment (emmadrigal) [reply tests/unit/r2i/tensorrt/engine.cc]: Some of the classes need some initialization in the other modules. These includes allows setting up the data, while only mocking the TensorRT functions

File: r2i/tensorrt/parameters.cc [hunk 48]
```diff
@@ -0,0 +1,231 @@
...
+  }
+
+Parameters::Parameters () :
+  parameter_map ({
+  /* Global parameters */
+  PARAM("version", "TensorRT version",
+        r2i::ParameterMeta::Flags::READ,
+        r2i::ParameterMeta::Type::STRING,
+        std::make_shared<VersionAccessor>(this)),
+  PARAM("batch_size", "Batch size, can't be larger the one the engine was created with",
```
Comment (rrcarlosrodriguez) [inline r2i/tensorrt/parameters.cc]: can't be larger than the one...

File: tests/unit/r2i/tensorrt/prediction.cc [hunk 49]
```diff
@@ -0,0 +1,110 @@
...
+ * The contents of this software are proprietary and confidential to RidgeRun,
+ * LLC.  No part of this program may be photocopied, reproduced or translated
+ * into another programming language without prior written consent of
+ * RidgeRun, LLC.  The user is free to modify the source code after obtaining
+ * a software license from RidgeRun.  All source code changes must be provided
+ * back to RidgeRun without any encumbrance.
+*/
+
+#include <r2i/r2i.h>
+#include <r2i/tensorrt/prediction.h>
```
Comment (rrcarlosrodriguez) [inline tests/unit/r2i/tensorrt/prediction.cc]: Separate includes, and alphabetic order

File: tests/unit/r2i/tensorrt/mockcudaengine.cc [hunk 50]
```diff
@@ -0,0 +1,315 @@
+bool fail_context = false;
```
Comment (rrcarlosrodriguez) [inline tests/unit/r2i/tensorrt/mockcudaengine.cc]: What about the header in this one?

File: configure.ac [hunk 51]
```diff
@@ -66,18 +71,21 @@ docs/uml/Makefile
 examples/Makefile
 examples/r2i/Makefile
 examples/r2i/ncsdk/Makefile
+examples/r2i/tensorrt/Makefile
```
Comment (michaelgruner) [inline configure.ac]: Alphabetical order

File: configure.ac [hunk 52]
```diff
@@ -66,18 +71,21 @@ docs/uml/Makefile
...
 r2i/tensorflow/Makefile
+r2i/tensorrt/Makefile
 r2i/tflite/Makefile
 r2inference-${RR_PACKAGE_VERSION}.pc:r2inference.pc.in
 tests/Makefile
 tests/acceptance/Makefile
 tests/unit/Makefile
 tests/unit/r2i/Makefile
 tests/unit/r2i/ncsdk/Makefile
+tests/unit/r2i/tensorrt/Makefile
```
Comment (michaelgruner) [inline configure.ac]: Alphabetical order

File: r2i/datatype.cc [hunk 53]
```diff
@@ -0,0 +1,58 @@
...
+#include "r2i/datatype.h"
+
+#include <string>
+#include <unordered_map>
+
+namespace r2i {
+
+static std::unordered_map<int, std::pair<const std::string, int>>
+data_type_descriptors ({
+  {DataType::Id::INT32, {"32 bit integer", 3}},
```
Comment (michaelgruner) [inline r2i/datatype.cc]: Is this 3 the size? If so, this should be 4.

File: r2i/ncsdk/frame.h [hunk 54]
```diff
@@ -30,6 +30,8 @@ class Frame : public IFrame {
 
   ImageFormat GetFormat () override;
 
+  virtual DataType GetDataType () override { return r2i::DataType::Id::UNKNOWN_DATATYPE; }
```
Comment (michaelgruner) [inline r2i/ncsdk/frame.h]: This is a float32

File: r2i/tensorflow/frame.h [hunk 55]
```diff
@@ -37,6 +37,8 @@ class Frame : public IFrame {
   std::shared_ptr<TF_Tensor> GetTensor (std::shared_ptr<TF_Graph> graph,
                                         TF_Operation *operation, RuntimeError &error);
 
+  virtual DataType GetDataType () override { return r2i::DataType::Id::UNKNOWN_DATATYPE; }
```
Comment (michaelgruner) [inline r2i/tensorflow/frame.h]: This is a float32

File: r2i/ncsdk/frame.h (hunk 54 above)
Comment (michaelgruner) [reply r2i/ncsdk/frame.h]: Dont implement it in the header.

File: r2i/tensorflow/frame.h (hunk 55 above)
Comment (michaelgruner) [reply r2i/tensorflow/frame.h]: Dont implement it in the header

File: r2i/tensorrt/Makefile.am [hunk 56]
```diff
@@ -0,0 +1,56 @@
...
+    frameworkfactory.cc          \
+    loader.cc                    \
+    model.cc                     \
+    parameters.cc                \
+    prediction.cc
+
+libtensorrt_la_CPPFLAGS =        \
+    $(RR_CPPFLAGS)               \
+    $(CODE_COVERAGE_CPPFLAGS)    \
+    -I$(CUDA_INCDIR)
```
Comment (michaelgruner) [inline r2i/tensorrt/Makefile.am]: Don't we have CUDA_CFLAGS?

File: r2i/tensorrt/engine.cc [hunk 57]
```diff
@@ -0,0 +1,185 @@
...
+ */
+
+#include "r2i/tensorrt/engine.h"
+#include "r2i/tensorrt/frame.h"
+#include "r2i/tensorrt/prediction.h"
+
+#include <cuda.h>
+#include <cuda_runtime.h>
+#include <NvInfer.h>
+#include "NvInferPlugin.h"
```
Comment (michaelgruner) [inline r2i/tensorrt/engine.cc]: This one is not needed.

File: r2i/tensorrt/engine.cc [hunk 58]
```diff
@@ -0,0 +1,185 @@
...
+      buffers.emplace_back (in_frame->GetData());
+    } else {
+
+      output_size = 1;
+      for (int d = 0; d < dims.nbDims; ++d) {
+        output_size *= dims.d[d];
+      }
+      output_size *= this->batch_size;
+
+      output_size *= sizeof(float);
```
Comment (michaelgruner) [inline r2i/tensorrt/engine.cc]: Use the data size, instead of the hardcoded value.

File: r2i/tensorrt/loader.cc [hunk 59]
```diff
@@ -0,0 +1,136 @@
...
+      case
+          Severity::kINFO:
+        tag = "INFO";
+        break;
+      case
+          Severity::kVERBOSE:
+        tag = "VERBOSE";
+        break;
+    }
+  }
```
Comment (michaelgruner) [inline r2i/tensorrt/loader.cc]: Leave a comment on why this function doesn't do anything.

File: r2i/tensorrt/model.h [hunk 60]
```diff
@@ -0,0 +1,50 @@
...
+ * back to RidgeRun without any encumbrance.
+*/
+
+#ifndef R2I_TENSORRT_MODEL_H
+#define R2I_TENSORRT_MODEL_H
+
+#include <iostream>
+#include <memory>
+#include <NvInfer.h>
+#include "NvInferPlugin.h"
```
Comment (michaelgruner) [inline r2i/tensorrt/model.h]: This is not needed.

File: r2i/tensorrt/prediction.cc [hunk 61]
```diff
@@ -0,0 +1,99 @@
...
+  error.Clean ();
+
+  if (nullptr == this->result_buffer) {
+    error.Set (RuntimeError::Code::INVALID_FRAMEWORK_PARAMETER,
+               "Prediction was not properly initialized");
+    return 0;
+  }
+
+  /* FIXME get correct datatype */
+  unsigned int n_results =  this->GetResultSize() / sizeof(float);
```
Comment (michaelgruner) [inline r2i/tensorrt/prediction.cc]: Don't use the hardcoded sizeof(float), use the data type size as provided by your DataType class.

File: r2i/tensorrt/prediction.cc [hunk 62]
```diff
@@ -0,0 +1,99 @@
...
+  if ( n_results < index ) {
+    error.Set (RuntimeError::Code::MEMORY_ERROR,
+               "Triying to access an non-existing index");
+    return 0;
+  }
+
+  float *fdata = static_cast<float *> (this->GetResultData ());
```
Comment (michaelgruner) [inline r2i/tensorrt/prediction.cc]: Same here.

File: r2i/tensorrt/Makefile.am (hunk 56 above)
Comment (emmadrigal) [reply r2i/tensorrt/Makefile.am]: yes, but they add some flags that appear to be meant for nvcc rather than g++>

**Pull request merged**